admin-dev/themes/new-theme/js/components/modal/progress-modal.ts

```diff
@@ -0,0 +1,514 @@
+    super.initContainer(params);
+    this.progressModal = new ProgressModalContainer(params);
+    this.errorModal = new ProgressModalErrorContainer(params);
+
+    this.modal.dialog.appendChild(this.progressModal.content);
+
+    this.initListeners(params);
+  }
+
+  public updateCount(doneCount: number)
+  {
+    this.updatePercentage(doneCount);
+  }
+
+  private updatePercentage(doneCount: number)
+  {
+    let progressBarDone = doneCount * 100 / this.total;
+    this.progressModal.progressDone.style.width = progressBarDone+'%';
+    this.progressModal.progressPercentage.innerHTML = progressBarDone.toFixed()+'%';
+  }
+
+  public addError(error: string)
+  {
+    this.errors.push(error);
+    let errorContent = document.createElement('p');
+    errorContent.classList.add(
+      'progress-modal-error',
+    );
+    errorContent.append(this.getErrorIcon());
+    errorContent.append(error);
+    this.errorModal.errorContainer.append(errorContent);
+    this.progressModal.switchToErrorButton.innerHTML = this.params.modalViewErrorLog.replace('%d', this.errors.length.toString());
+    this.progressModal.lastError.classList.remove('d-none');
+    this.progressModal.lastError.innerHTML = error;
+    if (!this.processStopped) {
+      this.progressModal.progressIconContainer.innerHTML = this.getErrorIcon().outerHTML;
+    }
+  }
+
+  /** Return 401 on error */
+  private initListeners(params: ProgressModalParams)
+  {
+    this.errorModal.downloadErrorsButton.addEventListener('click', () => {
+      let csvContent = 'data:text/csv;charset=utf-8,';
+      this.errors.forEach(function(error) {
+        csvContent += error + "\r\n";
+      });
+      let link = document.createElement('a');
+      link.href = encodeURI(csvContent);
+      link.download = 'Errors.csv';
+      link.click();
+    });
+
+    this.errorModal.switchButton.addEventListener('click', () => {
+      this.modal.dialog.removeChild(this.errorModal.content);
+      this.modal.dialog.appendChild(this.progressModal.content);
+    });
+
+    this.progressModal.switchToErrorButton.addEventListener('click', () => {
+      this.modal.dialog.removeChild(this.progressModal.content);
+      this.modal.dialog.appendChild(this.errorModal.content);
+    });
+
+    this.progressModal.stopProcessingButton.addEventListener('click', () => {
+      if (params.cancelCallback) {
+        params.cancelCallback();
+        this.stopProcess();
+      }
+    });
+  }
+
+  public finishProcess()
+  {
+    if (this.processStopped) {
+      return
+    }
+
+    this.replaceStopProcessButton();
+
+    if (!this.errors.length) {
+      this.progressModal.progressIconContainer.innerHTML = this.getCompleteIcon().outerHTML;
+    }
+    this.processStopped = true;
+  }
+
+  public stopProcess()
+  {
+    if (this.processStopped) {
+      return
+    }
+
+    this.replaceStopProcessButton();
+    this.progressModal.progressIconContainer.innerHTML = this.getStopIcon().outerHTML;
+    this.processStopped = true;
+  }
```

Comment:
You could use a private common methods for those two public ones since they are very similar:

```suggestion
  public completeProcess(): void
  {
    this.stopProcess(this.errors.length > 0 ? this.getErrorIcon() : this.getCompleteIcon());
  }

  public interruptProcess(): void {
    this.stopProcess(this.getStopIcon());
  }

  private stopProcess(progressIcon: HTMLElement): void
  {
    if (this.processStopped) {
      return
    }

    this.replaceStopProcessButton();
    this.progressModal.progressIconContainer.innerHTML = this.getStopIcon().outerHTML;
    this.processStopped = true;
  }
```